bundle/libraries/upload.go

```diff
@@ -0,0 +1,225 @@
+			file := v.MustMap()
+			sv, ok := file.GetByString("source")
+			if !ok {
+				return v, nil
+			}
+
+			source := sv.MustString()
+			libs[source] = append(libs[source], configLocation{
+				configPath: p.Append(dyn.Key("remote_path")),
+				location:   v.Location(),
+			})
+
+			return v, nil
+		})
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return libs, nil
+}
+
+func (u *upload) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
+	uploadPath, err := GetUploadBasePath(b)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// If the client is not initialized, initialize it
+	// We use client field in mutator to allow for mocking client in testing
+	if u.client == nil {
+		filer, err := GetFilerForLibraries(b.WorkspaceClient(), uploadPath)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		u.client = filer
+	}
+
+	var diags diag.Diagnostics
+
+	libs, err := collectLocalLibraries(b)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	errs, errCtx := errgroup.WithContext(ctx)
+	errs.SetLimit(maxFilesRequestsInFlight)
+
+	for source := range libs {
+		errs.Go(func() error {
+			return UploadFile(errCtx, source, u.client)
+		})
+	}
+
+	if err := errs.Wait(); err != nil {
+		return diag.FromErr(err)
+	}
+
+	// Update all the config paths to point to the uploaded location
+	for source, locations := range libs {
+		err = b.Config.Mutate(func(v dyn.Value) (dyn.Value, error) {
+			remotePath := path.Join(uploadPath, filepath.Base(source))
+
+			// If the remote path does not start with /Workspace or /Volumes, prepend /Workspace
+			if !strings.HasPrefix(remotePath, "/Workspace") && !strings.HasPrefix(remotePath, "/Volumes") {
+				remotePath = "/Workspace" + remotePath
+			}
+			for _, location := range locations {
+				v, err = dyn.SetByPath(v, location.configPath, dyn.NewValue(remotePath, []dyn.Location{location.location}))
+				if err != nil {
+					return v, err
+				}
+			}
+
+			return v, nil
+		})
+
+		if err != nil {
+			diags = diags.Extend(diag.FromErr(err))
+		}
+	}
+
+	return diags
+}
+
+func (u *upload) Name() string {
+	return "libraries.Upload"
+}
+
+func GetFilerForLibraries(w *databricks.WorkspaceClient, uploadPath string) (filer.Filer, error) {
+	if isVolumesPath(uploadPath) {
+		return filer.NewFilesClient(w, uploadPath)
+	}
+	return filer.NewWorkspaceFilesClient(w, uploadPath)
+}
+
+func isVolumesPath(path string) bool {
+	return strings.HasPrefix(path, "/Volumes/")
+}
+
+// Function to upload file (a library, artifact and etc) to Workspace or UC volume
+func UploadFile(ctx context.Context, file string, client filer.Filer) error {
+	filename := filepath.Base(file)
+	cmdio.LogString(ctx, fmt.Sprintf("Uploading %s...", filename))
+
+	raw, err := os.ReadFile(file)
+	if err != nil {
+		return fmt.Errorf("unable to read %s: %w", file, errors.Unwrap(err))
+	}
+
+	err = client.Write(ctx, filename, bytes.NewReader(raw), filer.OverwriteIfExists, filer.CreateParentDirectories)
```

Comment:
We don't need to buffer the file in memory.

With `os.Open` we get a file reader that can be passed to `client.Write`.